### PR TITLE
Janek/improve protocol interface

### PR DESCRIPTION
**Pull request by jqmp**
See commits for details.

### CODE REVIEW

File: bionic/protocols.py [hunk 1]
```diff
@@ -48,33 +48,69 @@ def check_is_like_protocol(obj):
...
 
     def read(self, path):
+        """Deserializes an object from the pathlib path ``path``, and returns it."""
         raise NotImplementedError()
 
+    def read_with_extension(self, path, extension):
+        """
+        Deserializes an object from the pathlib path ``path``, and returns it.
+
+        May use the file extension ``extension`` to determine what deserialization
```
Comment (namanjain) [inline bionic/protocols.py:108]: nit: is file extension ``extension`` a typo?

File: bionic/protocols.py [hunk 2]
```diff
@@ -86,15 +122,22 @@ def read(self, path):
...
+        """
+        Converts a valid value into a unique hash which can be used as a filename.
+        """
         if type(value) in self.SIMPLE_TYPES:
             return tokenization.tokenize(value)
         else:
             return tokenization.tokenize(value, self._write_to_bytes)
 
     def tokenize_file(self, path):
+        """Like ``tokenize``, but operates on a serialized value."""
```
Comment (namanjain) [inline bionic/protocols.py:134]: Can we be more specific here that the serialized value is coming from the contents of the path file (or dir)?

Comment (namanjain) [reply bionic/protocols.py:134]: Related to this, even I have trouble remembering if this path is a pathlib path. I remember seeing [typeguard](https://github.com/agronholm/typeguard) before for this, how do you feel about using for at least paths in bionic?

File: bionic/protocols.py (hunk 1 above)
Comment (jqmp) [reply bionic/protocols.py:108]: No, I was trying to use the same pattern as "pathlib path ``path``" above. If I just wrote "the file ``extension``" , that would sound to me like ``extension`` was a file.

**Comment by jqmp** [comment]
NOTE: for now, let's plan to avoid to merging this until https://github.com/square/bionic/pull/114 is merged.

**Comment by jqmp** [comment]
Based on offline conversations, #114 isn't going get merged for a few more days, so I'm going to go ahead and merge this PR.

**Pull request merged**